### PR TITLE
Search: Use aliased Preact imports

### PR DESCRIPTION
**Pull request by jsnmoon**
#### Changes proposed in this Pull Request:
In preparation for using our Instant Search application within Gutenberg, we'll rely on using aliased Preact imports using our Webpack configuration. 

(In other words, our codebase will read as though it's importing React even though it's actually importing Preact.)

#### Jetpack product discussion
Necessary for our future customization efforts, see p1HpG7-c6N-p2.

#### Does this pull request change what data or activity we track or use?
No.

#### Testing instructions:
* Apply this change to your Jetpack site with a Search subscription.
* Ensure that the search application builds as expected.
* Ensure that the search interface works as expected.

### CODE REVIEW

**Comment by bluefuton** [comment]
Looks good, but let's add some docs to help future us.

Can we add a 'React and Preact' section to the README explaining the background and how this is set up?

**Comment by jsnmoon** [comment]
> Looks good, but let's add some docs to help future us.
> 
> Can we add a 'React and Preact' section to the README explaining the background and how this is set up? (@bluefuton)

Done!

File: projects/plugins/jetpack/modules/search/instant-search/README.md [hunk 1]
```diff
@@ -11,22 +11,28 @@ It enables the user to quickly jump into a full-page search experience with filt
...
+> # alternatively, `pnpm build` or `pnpm watch`
+> # verify built assets via `ls _inc/build/instant-search/`
 ```
 
-### Testing Instructions
+### Webpack Architecture
+
+While the Instant Search application is primarily designed to be run as a Preact application, we use React imports in our code to ensure interoperability in React contexts, like in the Gutenberg editor.
+
+Using our Webpack configuration, we alias all `react` and `react-dom` to `preact/compat` to ensure that we ship Preact, not React, in our main bundle.
```
Comment (bluefuton) [inline projects/plugins/jetpack/modules/search/instant-search/README.md:33]: Excellent! 👍

**Comment by bluefuton** [comment]
Oddly this seems to break for me in the Twenty Twenty One theme, with a seemingly unrelated error:

Screen Shot 2021-06-04 at 11 07 39

This doesn't happen on Jetpack `master`. Can you reproduce it too?

My build command was `pnpm build && pnpm build-search -- --watch`.

**Comment by jsnmoon** [comment]
> Oddly this seems to break for me in the Twenty Twenty One theme, with a seemingly unrelated error:
> 
> Screen Shot 2021-06-04 at 11 07 39
> 
> This doesn't happen on Jetpack `master`. Can you reproduce it too?
> 
> My build command was `pnpm build && pnpm build-search -- --watch`. (@bluefuton)

I discovered that this was being caused by using `preact/compat`'s `{ render }` function in `index.jsx`, which behaves differently than `preact`'s `{ render }` function.

Since `index.jsx` is never used in a React context, I went ahead and reverted this change. Added the requisite code comment explaining the rationale as well.

**Comment by jsnmoon** [comment]
Unfortunately, this breaks two of our existing Jest tests due to Jest using its own import resolutions. I'm currently working on a fix.

File: projects/plugins/jetpack/tools/webpack.config.search.js [hunk 2]
```diff
@@ -105,6 +105,9 @@ module.exports = {
 			requestToExternal,
 			requestToHandle: defaultRequestToHandle,
 		} ),
+		new webpack.ProvidePlugin( {
```
Comment (kangzj) [inline projects/plugins/jetpack/tools/webpack.config.search.js:108]: Just curious where is `h` used?

**Review by kangzj — DISMISSED** [review]
It works well and looks good to me 👍

File: projects/plugins/jetpack/tools/webpack.config.search.js (hunk 2 above)
Comment (jsnmoon) [reply projects/plugins/jetpack/tools/webpack.config.search.js]: Good catch, this code is actually no longer necessary. For context, `preact.h` corresponds to `react.createElement` (and stands for [hyperscript](https://jasonformat.com/wtf-is-jsx/)).

**Review by jeherve — APPROVED** [review]
This looks good to me. 👍

**Comment by jsnmoon** [comment]
Thanks for merging, @sdixon194!

🚢 r227035-wpcom

**Pull request merged**